### PR TITLE
Dialog and button enhancements

### LINKED ISSUE #210: Revisit Button theme structure
Current Button theme logic: fg and bg colors are determined by `intent` and `appearance`…

If `appearance` is `default` or `minimal`, the color determined by `intent` is used for the foreground.  `base.colors.defaultBg` is used for the background in the `default` case, while background is transparent in the `minimal` case.

If `appearance` is `prominent`, then the color determined by `intent` is used for the _background_, and then the foreground color is algorithmically chosen from `base.colors.lightText` or `base.colors.darkText`.

And then you can always override any of the above via `color` and/or `bg` props.

@jfrankl Reported on some confusion using that model:
> I’m a little confused about what I was doing didn’t work, though, in terms of how I’d expect the theme to be structured. I’d expect a button created with no specified `appearance` or `intent` would tap into the base values. I wonder if there is a more clear way to organize this part of the theme.

### LINKED ISSUE #211: Dialog: Overlay color prop and theme prop

### LINKED ISSUE #212: Dialog: Support custom transitions

### LINKED ISSUE #213: Revisit Dialog focus treatment
When a Dialog opens, the `role="dialog"` element automatically gains focus, and so has an outline, eg:

In practice, it's probably best for consumers to programmatically give focus to an interactive element within the Dialog when it opens. But we should consider revisiting the Dialog focus treatment for cases when that isn't done. 

For a11y's sake, we should retain some indication of focus, but it can be opinionated.

Note: the Dialog focus can already be disabled by passing `shouldFocusAfterRender={false}` to the `<Dialog>`, but this leaves the focus on the button that triggered the dialog, which is outside the dialog boundaries and so creates havoc when attempting to navigate the open dialog by keyboard.

### PR DESCRIPTION
**Pull request by lederer**
## Overview

Button
- Theme structure changed so default colors are now configured in `colors`.

Dialog
- Overlay color and opacity can now be themed or propped.
- Custom transitions can now be themed.
- z-index can now be propped.
- Focus ring color can now be themed, and has a default value.

### Checklist

- [x] Relevant documentation pages have been created or updated
- [x] Description of PR is in an appropriate section of the changelog and grouped with similar changes if possible
- [ ] Relevant status has been updated on the status page

### Upgrade instructions

If there are any of the following in this PR, provide proper instructions on how to upgrade:

- [x] Changes to the prop names or types of existing components
- [ ] Changes in intended behavior of existing component props
- [x] Changes in the theme file's structure
- [ ] A required version bump to a non-dev dependency of the project

Button theme changes:
- `intents.colors.default` is now `base.colors.color`
- `base.colors.defaultBg` is now `base.colors.bg`
- `base.colors.defaultBorder` is now `base.colors.border`
- `base.colors.darkText` is now `base.colors.colorDark`
- `base.colors.lightText` is now `base.colors.colorLight`
- `base.hover.colors.defaultBg` is now `base.colors.bgHover`
- All other `base.hover.*` properties deprecated, and `base.hover` removed.

Dialog prop changes:
- `opacity` should now affect dialog opacity, not overlay opacity
- Added `overlayOpacity`
- Added `overlayColor`
- Added `zIndex`
- Deprecated: `portalClassName`, `overlayClassName`, `className`, `bodyOpenClassName`, `htmlOpenClassName`

Dialog theme changes:
- Added `colors.overlay`
- Added `colors.focus`
- Added `zIndices.zIndex`
- Added `opacities.overlay`
- Added `durations.closeTimeoutMS`
- Added `transitions.overlay.{base, afterOpen, beforeClose}`
- Added `transitions.dialog.{base, afterOpen, beforeClose}`

### Notes

Tweaked `replaceThemeRefs()` in utils.js to help prevent call stack errors.

Hard coded Dialog classnames, for simplicity.

Hid Field, ToggleField, and NavBar docs pages. They're to be deprecated.

## Testing Instructions

- Fiddle with new Button theme values. Confirm works as expected.
- Fiddle with new Dialog props and theme values. Confirm works as expected.

Closes #210 
Closes #211 
Closes #212 
Closes #213

### CODE REVIEW

File: packages/core/components/dialog/index.js [hunk 1]
```diff
@@ -99,40 +119,49 @@ const Dialog = ({
...
 
     /******************
      * Dialog
      */
-    .${className.base} {
-      transform: translateY(200px);
-      transition: transform ${closeTimeoutMS}ms ease-in-out;
+    .${CLASSES.DIALOG.base} {
+      transition-duration: ${closeTimeoutMS}ms;
+      transition-function: ease-in-out;
```
Comment (designmatty) [inline packages/core/components/dialog/index.js]: should be `transition-timing-function`

File: packages/core/components/dialog/index.js [hunk 2]
```diff
@@ -99,40 +119,49 @@ const Dialog = ({
...
 
     /******************
      * Overlay
      */
-    .${overlayClassName.base} {
-      opacity: 0;
-      transition: opacity ${closeTimeoutMS}ms ease-in-out;
+    .${CLASSES.OVERLAY.base} {
+      transition-duration: ${closeTimeoutMS}ms;
+      transition-function: ease-in-out;
```
Comment (designmatty) [inline packages/core/components/dialog/index.js]: should be: `transition-timing-function`

**Review by designmatty — CHANGES_REQUESTED** [review]
Some weird alignment issues when you have more copy in the modal.

Screen Shot 2019-08-12 at 3 12 04 PM

**Comment by lederer** [comment]
Nice catches. Ready for another look.

**Review by designmatty — APPROVED** [review]
Looks good now. Thanks for making those changes.

**Pull request merged**